### DISCUSSION TITLE
Megathread: the name "sudo"?

### DISCUSSION THREAD

**Discussion by poolpOrg** [microsoft/sudo#23] [General]
You _really_ couldn't come up with a different name to avoid the confusion when writing a software of your own that did what sudo does but for Windows ?

It's not like sudo is already a software that's widespread and yours is not even compatible with...

**Comment by sirredbeard** [comment]
As someone who regularly works on Linux, BSD, and Windows, I appreciate that Microsoft supports many common *NIX commands in PowerShell: `ls`, `curl`, `ssh`, `cat`, `clear`, `cp`, `echo`, `kill`, `mv`, `ps`, `rmdir`, and now `sudo`, all _just work_. It significantly reduces the cognitive load of context switching and allows me to focus on getting the task at hand done. Developers, system administrators, engineers, and power users are likely familiar with `sudo` from work on other operating systems. Adopting `sudo` will be easier and make work more efficient for them.

**Reply by nukeop** [reply]
But it doesn't support those commands. Those are just lookalikes piggybacking on names of superior programs to help build some false trust. They are illegitimately exploiting existing brands in order to confuse users about their origins. It reduces the cognitive load of context switching in the same sense that pickpocketing reduces the financial burden for the thief.

If I made a Linux distro, called it Windows, and added some lookalike Windows tools, I doubt Microsoft would appreciate that my distro supports their tools.

**Reply by jimscard** [reply]
Windows is a trademark. You couldn't call your theoretical distro "Windows" because of that.

On the other hand, computer command structures are neither trademarkable nor copyrightable.  And there already are Linux distros that are lookalike to Windows. For example, the 'undercover' mode in Kali Linux, not to mention actual emulators like Wine, Crossover Linux....

**Reply by nukeop** [reply]
It's not a structure, it's a name of a program which is comparable to a brand. A de facto trademark enjoys the same protections.

**Comment by poolpOrg** [comment]
> As someone who regularly works on Linux, BSD, and Windows, I appreciate that Microsoft supports many common *NIX commands in PowerShell: `ls`, `curl`, `ssh`, `cat`, `clear`, `cp`, `echo`, `kill`, `mv`, `ps`, `rmdir`, and now `sudo`, all _just work_. It significantly reduces the cognitive load of context switching and allows me to focus on getting the task at hand done. Developers, system administrators, engineers, and power users are likely familiar with `sudo` from work on other operating systems. Adopting `sudo` will be easier and make work more efficient for them.

Except that Microsoft's sudo is not sudo, this is akin to me creating a text editor project and naming it Visual Studio Code or writing a window manager and naming it Windows.

As far as I know, *NIX commands in PowerShell are mimic-ing their unix counterparts, but this is not doing that, literally from the README file:

```
This project is not a fork of the Linux sudo project, nor is it a port of the Linux sudo project.
```

**Reply by Jaykul** [reply]
Technically, none of the "*NIX commands" in PowerShell are _duplicates_ of their counterparts at all. 

They superficially perform the same function, and they have the same syntax for the _most common_ uses, but they have different calling syntax, different pipeline syntax, different parameters, different output. This is a lot like that, actually.

The syntax is the same for the most common uses, and you can _guess_ what command to run, so you have a head start -- and importantly, you can get help for them with `man` (which of course, is also nothing like the real man).

**Reply by Krys87** [reply]
Most if not all of these nix commands are aliases to pwsh `cmdlets`

```
PS C:\> alias clear

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           clear -> Clear-Host

PS C:\> alias ls

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           ls -> Get-ChildItem

PS C:\> alias cat

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           cat -> Get-Content
```

They do a similar job but they arent the same in functionality

**Reply by poolpOrg** [reply]
None of these [POSIX commands](https://en.wikipedia.org/wiki/List_of_POSIX_commands) are an opensource project whose name is being used by ... another opensource project, but nevermind this doesn't seem to bother anyone 🙄

**Reply by jimscard** [reply]
You can't copyright command structures.. There's nothing wrong or improper about implementing a command on your operating system that has similar functionality and takes similar arguments as something someone else did.  There would be a problem if you reused someone's code without permission, or someone else's Trademark -- but 'sudo' isn't a trademarked product name, it's the actual command.
On a side note, Microsoft was at one point in time the highest-volume provider of Unix systems.  The "su" in "sudo" is a reference to the Unix superuser command (`su`), so one could argue that "sudo" caused confusion with su .. :)

**Comment by sirredbeard** [comment]
> *NIX commands in PowerShell are mimic-ing their unix counterparts

Yes, and sometimes the options for those commands differ between their *NIX counterparts and PowerShell implementations, but most of the time they just work.

Another example would be even though Windows has implemented `ssh` based on OpenSSH, there are still some behavioral differences between `ssh` on *NIX and `ssh` on Windows you will run into in some edge cases.

Along those lines, `sudo` for Windows mimics `sudo` on *NIX but has a Windows-specific implementation owing to its distinct permissions system. 

The goal overall goal, whether it be PowerShell aliases, re-implementation, or integrations of upstream open source tools is similar functionality where possible, subject to differences between the distinct OS lineages, to reduce the friction in day-to-day use for users.

**Reply by pv2b** [reply]
The difference is that "ssh" is still OpenSSH, and while there are some differences, it's not a fundamentally different program.

**Comment by omnivagant** [comment]
Don't you have aliases or something in DOS? My `ls` is `eza -F` and my `cat` is `bat --wrap character -p` and so on.

Regardless of that, a project can be named something other than the components it is providing, like the `ssh` client is provided by, for example, the OpenSSH and Dropbear projects. Why not call it "MS sudo" or "sudo4win"? And, you know, all the cool kids switched from `sudo` years ago, so you could go with "Windoas".

Also, sudo isn't a kernel project, don't drag Linux into this.

**Reply by jimscard** [reply]
Well, it's a feature in some version of Windows, so it's called "Windows".  The command structure is the same at some level as the tool that the Sudo Project releases for Linux, no doubt with changes and extensions that make sense on Microsoft's OS.  This isn't the first release of sudo functionality for Windows, by the way.  A number of years ago, everyone panicked when Microsoft got a patent on their version of the sudo functionality (run a program as superuser; require authentication by an authorized user to do so).  It was a GUI based tool, you may have heard of it -- User Account Control.

**Reply by omnivagant** [reply]
My comment was mostly meant to be humourous. I honestly don't care what the Microsoft project is called but tried to suggest that it didn't _need_ to have the same name as the command it provided. I included a bunch of nerdy references to try to come across as silly.

Of course I understand that the point is to have a command, binary or alias (or link or shortcut or what you call it), available with the same name and similar functionality in order to be ergonomical for people comming from and/or being used to Unix®-like environments.

However, I don't think that [the sudo project](https://github.com/sudo-project/sudo) should be referenced to as "the Linux `sudo` project", as the sudo project predates Linux by over a decade and is not limited to Linux-based systems and is, frankly, not a Linux project.

**Comment by joshtriplett** [comment]
There's a long history of Open Source OSes providing reimplementations of existing commands; there are many implementations of `ls`, `cp`, `vi`, and similar. I don't see anything wrong with providing another implementation of `sudo`; I think that's a great step towards being compatible with the numerous instructions across the web that show using `sudo xyz` to run something as root.

That said, since this is using the name `sudo`, it should (eventually) aim to be compatible with at least the most common options of `sudo`, such as (eventually) supporting options like `-H`, `-u`, `-i` (possibly just ignored for compatibility), `-s`, `-E`, and `-D`.

Using the name `sudo` and being largely compatible with the options of `sudo` seems great. Using the name `sudo` and *not* being compatible with the most common `sudo` options would be a repeat of what happened with `curl`.

Supporting `-u`: https://github.com/microsoft/sudo/issues/3

Supporting `-H`: https://github.com/microsoft/sudo/issues/19

Supporting `-s`: https://github.com/microsoft/sudo/issues/20

Supporting `-E`: https://github.com/microsoft/sudo/issues/21

Supporting `-D`: https://github.com/microsoft/sudo/issues/22

**Reply by pv2b** [reply]
The major difference is that something like "ls", "cp" and "vi" all attempt to be extremely close in functionality across platforms. "sudo" will neccessarilly be fundamentally differenct because the security model on Windows is fundamentally different to the one in Linux and Unix.

Nobody complains about GNU and BSD ls being different because they are very similar in functionality, and attempt to mimic each other as fas as possible. That won't be the case with a Windows sudo.

**Reply by jimscard** [reply]
Well, GNU attempts to mimic Unix. BSD *is* Unix.  The point is valid, however, about command structures being as similar as they can be, which benefits the many of us who have to work in a wide variety of operating systems on a daily basis.
There's always going to be differences -- some options aren't going to get implemented for one reason or another, and there'll be extensions on one platform that don't make sense for another.

**Reply by joshtriplett** [reply]
People *absolutely* complain about the differences between GNU and BSD and macOS implementations of tools. For instance, some tools allow options after arguments and some don't; some tools accept --help and some don't...

I don't think sudo has to be a perfect clone to be reasonable, as long as it's command-line compatible with the vast majority of the functionality people use.

**Reply by pv2b** [reply]
Sure, there are minor differences between GNU and BSD tools, but they are minor. The real sudo and "Sudo For Windows" are ver different. It's like aliasing vi to notepad and saying "actually, it's the same thing, both are text editors".

**Reply by joshtriplett** [reply]
Any sudo install that uses PAM could be configured to prompt the user graphically as well. In practice, the differences are small enough that when people read a series of instructions written with Linux in mind that involve `sudo`, the sudo command will likely work for the purposes of those instructions.

**Comment by lrvl** [comment]
I appreciate the effort to bring the functionality of a popular command to Windows, with the potential to bridge platform gaps and enhance user experience. However, this recreated version lacks significant compatibility and features, which raises concerns about integration with the existing ecosystem.

Let's encourage the development team to closely collaborate with the community. This collaboration could lead to a deeper understanding of the necessary features and compatibility standards to ensure the tool aligns more closely with users' expectations and the requirements of the matured broader ecosystems like Linux and BSD based, which are the defacto free, open and transparent operating systems.

Supporting open standards and constructive dialogue is vital. We look forward to efforts that strengthen, rather than **fragment**, our shared technological landscape.

**Comment by joadoumie** [comment]
We fully understand that there are differences between Sudo for Windows and sudo on other operating systems. We’ve already taken the time to listen to folks on what makes the most sense for them in their workflows and naming this project Sudo for Windows and shipping `sudo.exe` was the best way to make sure that folks can elevate processes in Windows similarly to how they do so elsewhere (and they can do that from different shells). It's great for folks who context-switch a lot or may not even know that sudo has not existed on Windows before this project. As mentioned previously in the thread, *NIX command naming conventions already exist in PowerShell that are designed to help users feel at home regardless of where they develop because they can use the same commands on different operating systems and expect similar behaviors. Of course, Windows is fundamentally different than *NIX operating systems and so `sudo.exe` will not behave the same.

This has been a feature that has been highly asked for long before I joined Microsoft, and this was the best way that our team identified to bring the feature to you all. I'm so excited about the community feedback and responses and we want to make this thing as useful as possible for everyone. 

If there are behaviors that are missing that you want to see here, please 100% let us know the use case by creating an issue or give an existing issue a thumbs up so we can prioritize it and make Sudo for Windows as great as possible.

**Reply by nukeop** [reply]
Is anything preventing you from bringing that feature to windows without falsely associating it with an established program that has nothing to do with your implementation? Can I fork Chromium, and release it as Edge for Linux?

**Reply by Brynyard** [reply]
The problem is that you say you've listened, yet still you continually ignore even the most basic concepts of a TTY.

The feature I know has been highly requested is to allow the TTY to be operated through T. That elevation also needs GUI access has been the fundamental problem, and there's already plenty of scripts/snippets/etc that allows one to launch a GUI requesting elevation, but again, GUI.

So shipping a "sudo" tool that doesn't doesn't let a terminal command run in the terminal is at best misleading.

**Reply by jimscard** [reply]
Nonsense. Like other commands that are universally implemented, SuperUser Do, aka sudo, functionality has to be adapted to make sense in the target operating system's paradigm.

Which Microsoft has done before - like when they implemented the functionality as UAC, and there was lots of noise from uniformed folks who thought Microsoft had patented sudo, the functionality, rather than UAC, the implementation.

People really need to stop beating this dead horse. "sudo" the word is not a trademark, and could not be, as computer commands and computer languages aren't trademarkable. Even if they were, the open license eliminates that route.  Additionally, they're not copyrightable - and with the license, anyone is free to release their own sudo with or without code from the sudo project's implementation.

OTOH, implementing the sudo command structure on Windows as appropriate to the windows OS has a clear community value.

**Reply by nukeop** [reply]
You're gonna love my new universally implemented utility, the Linux Edge

**Reply by Brynyard** [reply]
@jimscard I'm sorry to break this to you, but words have meanings, and you can't just redefine them as you like with excuses like having to "adapted" to a given targets operating systems "paradigms" (unless you're expecting people to just bend over backwards because it's Microsoft who should decide what words should mean).

Unless they add other ways of authenticating, they haven't added any new functionality. RunAs has existed for a long time, and relabeling some script to sudo doesn't _make_ it sudo.

**Comment by ZeyadMoustafaKamal** [comment]
To be honest as one who uses Windows all the time. I think that working on something that makes us able to do normal things faster should be the priority. I am not talking about some wrapper like this one that will prompt you to give the admin permission to the application that needs it. I mean something that changes windows from the bottom.

I thank you for what you have done. But don't you think that if we added a better sudo that works exactly like the one in Linux would be better??

And another thing that I want to say about installing important software for us. Have you tried to install Jenkins or Rust on Windows?? if not please try to do it. It will require a lot and a lot of things to do. For Rust, you will have to install something related to Visual Studio and about 1 GB as I remember. Just can you try to change Windows so it will just require me one step to install the software or to do something?

**Reply by zadjii-msft** [reply]
Thanks for the feedback, but this isn't really the repo to have that kind of discussion on. We're gonna keep things tightly focused on specifically Sudo for Windows, and be quite liberal with marking other discussion as off-topic. Thanks!

**Comment by pv2b** [comment]
You should be aware that naming your project "sudo" will cause bugs to be mis-reported to the "real" sudo project, and cause an excess of work for this project, as well as user confusion when they expect the two tools to be the same if they're named the same.

Please don't repeat the mistakes or aliasing curl and wget to sudo, change this before it's too late, and changing this would be a breaking change. The opportunity to change this is now, later will be too late.

That said, I think adding this type of command is great in principle, it just shouldn't have this name.

**Reply by pa-philos** [reply]
I would like to agree that probably would be better to name something else, but I myself use a different terminal (CMDER) for Windows just because I have most of the Linux common used commands available.

**Reply by chnhkk** [reply]
I agree. If you really want to use "sudo", you can just set an alias.

**Comment by IDisposable** [comment]
Can you please fix the windows installation so that when SUDO is disabled in developer settings, it doesn't create a placeholder `sudo.exe` which breaks our GSudo installations every time we run major Windows Updates? The installed sudo.exe is completely useless and merely pops up dialog showing where to enable it (RUDE).

No means No.

![image](https://github.com/user-attachments/assets/be659329-2084-48da-a753-8f83b3ebf914)

![image](https://github.com/user-attachments/assets/2537fb01-5dd3-4be4-9e83-6dd5873293c0)

![image](https://github.com/user-attachments/assets/63c0b262-f935-4fbd-a007-f43df80e4fe1)

**Reply by IDisposable** [reply]
p.s. to fix this after a windows install, run `gsudo config PathPrecedence True` to push the GSudo back to the front of the PATH
